# Purpose of Dataset Selected

The purpose of our project is to determine the reliability of an airline based on their prior on-time statistics spanning data for one year.

The following dataset is downloaded from the bureau of transportation website and contains metrics for domestic flights by major carriers.  It contains on-time arrival data such as departure and arrival delays, origin and destination airports, flight numbers, scheduled and actual departure and arrival times, cancelled or diverted flights, taxi-out and taxi-in times, air time, and non-stop distance.

For the purpose of this study, we downloaded data spanning March 2016 to February 2017.  There are 6,032,365 total records and 98 attributes in total.

# Business Understanding

In order to gauge whether the flight is reliable, we will use the following dataset features:  Departure delay, Arrival Delay, 
Cancelled, and Diverted.  

A reliable airline is defined as one where, if the flight is delayed/diverted then, departure delay is 
less than or equal to 60 minutes and arrival delay is less than or equal to 30 minutes.  In addition, the flight cancellation
rate should be <=30% for a given month.  

The data in the present dataset includes features such as weather delay, CarrierDelay,NASDelay, SecurityDelay and 
LateAircraftDelay timings in minutes.  One potential approach to predict the future reliability of the
flight could be to do cluster analysis for each airline by month based on the above delay conditions and predict future delays.
We can do stratified cross validation there forth and do Precision-Recall curve analysis to classify "positive" class outcomes.

# Data Meaning Type

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_colwidth', -1)
pd.set_option('display.max_columns',999)  # made all columns appear in the output

In [2]:

import pandas as pd
# Read the csv files by year and month, read only those fields we need (shown in Lab1 project)
fields = ['Month','DayofMonth','Carrier','Origin','Dest', 'DepDelayMinutes','DepTimeBlk' ,'TaxiOut','TaxiIn','ArrTimeBlk','ArrDelayMinutes', 'ArrTimeBlk', 

'Cancelled','CancellationCode','Diverted','ActualElapsedTime','DistanceGroup','CarrierDelay','WeatherDelay','NASDelay','SecurityDelay','LateAircraftDelay','Diverted']
# Array specifies the fields from the csv file to be read into the dataframe

# DataFrame for March 2016
df1 = pd.read_csv('2016_3.csv', skipinitialspace=True, usecols=fields)

# DataFrame for April 2016
df2 = pd.read_csv('2016_4.csv', skipinitialspace=True, usecols=fields)

# DataFrame for May 2016
df3 = pd.read_csv('2016_5.csv', skipinitialspace=True, usecols=fields)

# DataFrame for June 2016
df4 = pd.read_csv('2016_6.csv', skipinitialspace=True, usecols=fields)

# DataFrame for July 2016
df5 = pd.read_csv('2016_7.csv', skipinitialspace=True, usecols=fields)

# DataFrame for August 2016
df6 = pd.read_csv('2016_8.csv', skipinitialspace=True, usecols=fields)

# DataFrame for September 2016
df7 = pd.read_csv('2016_9.csv', skipinitialspace=True, usecols=fields)

# DataFrame for October 2016
df8 = pd.read_csv('2016_10.csv', skipinitialspace=True, usecols=fields)

# DataFrame for November 2016
df9 = pd.read_csv('2016_11.csv', skipinitialspace=True, usecols=fields)

# DataFrame for December 2016
df10 = pd.read_csv('2016_12.csv', skipinitialspace=True, usecols=fields)

# DataFrame for January 2017
df11 = pd.read_csv('2017_1.csv', skipinitialspace=True, usecols=fields)

# DataFrame for February 2017
df12 = pd.read_csv('2017_2.csv', skipinitialspace=True, usecols=fields)


C:\Users\03307966\AppData\Local\Continuum\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2717: DtypeWarning: Columns (48) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
#df2.head()
df11.head()

Month  DayofMonth Carrier Origin Dest  DepDelayMinutes DepTimeBlk  TaxiOut  \
0  1      17          AA      CLT    PHX  0.0              1600-1659  17.0      
1  1      18          AA      CLT    PHX  0.0              1600-1659  13.0      
2  1      19          AA      CLT    PHX  0.0              1600-1659  17.0      
3  1      20          AA      CLT    PHX  37.0             1600-1659  18.0      
4  1      21          AA      CLT    PHX  13.0             1600-1659  17.0      

   TaxiIn  ArrDelayMinutes ArrTimeBlk  Cancelled CancellationCode  Diverted  \
0  5.0     0.0              1800-1859  0.0        NaN              0.0        
1  6.0     0.0              1800-1859  0.0        NaN              0.0        
2  2.0     0.0              1800-1859  0.0        NaN              0.0        
3  3.0     33.0             1800-1859  0.0        NaN              0.0        
4  4.0     2.0              1800-1859  0.0        NaN              0.0        

   ActualElapsedTime  DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  \
0  266.0              8             NaN           NaN           NaN         
1  247.0              8             NaN           NaN           NaN         
2  255.0              8             NaN           NaN           NaN         
3  273.0              8              33.0          0.0           0.0        
4  266.0              8             NaN           NaN           NaN         

   SecurityDelay  LateAircraftDelay  
0 NaN            NaN                 
1 NaN            NaN                 
2 NaN            NaN                 
3  0.0            0.0                
4 NaN            NaN

#### the output of the month should match the month being read in - it does

In [4]:
# List of data frames holding the individual monthly data
dataframes = [df1, df2, df3, df4, df5, df6, df7, df8, df9, df10, df11, df12] # to create sampled.csv of all rows

# Initialized the dataset which will hold 10% of samples per month
df = pd.DataFrame()

# Loop through each of the data frames
for dataframe in dataframes:
    df = pd.concat([df, dataframe.sample(frac=0.1)])  #used 10% OF EACH file or month
# Save the sample data to file
file_name = 'sampled.csv'
df.to_csv(file_name, sep=',', encoding='utf-8')

Now that we have sampled the dataset, lets take a look at the data and gather some basic information about this dataset. 

In [5]:
# Read the sampled csv file
#df_sampled = pd.read_csv('sampled.csv', index_col=None)
#df_sampled = pd.read_csv('sampled.csv', index_col=-1)  # lateaircraft delay is on left of unnamed:0
df_sampled = pd.read_csv('sampled.csv', index_col=0)  # has col with numbers on left of month

df_sampled.head()

Month  DayofMonth Carrier Origin Dest  DepDelayMinutes DepTimeBlk  \
28459   3      18          AA      MIA    IAH  76.0             1700-1759   
193274  3      19          DL      ATL    RDU  0.0              1900-1959   
183449  3      27          DL      CVG    LAX  0.0              1600-1659   
157662  3      18          DL      MCI    DTW  0.0              1700-1759   
304586  3      18          WN      LAS    BWI  7.0              0900-0959   

        TaxiOut  TaxiIn  ArrDelayMinutes ArrTimeBlk  Cancelled  \
28459   31.0     10.0    91.0             1900-1959  0.0         
193274  20.0     3.0     0.0              2100-2159  0.0         
183449  10.0     20.0    0.0              1700-1759  0.0         
157662  12.0     19.0    0.0              2000-2059  0.0         
304586  15.0     3.0     0.0              1600-1659  0.0         

       CancellationCode  Diverted  ActualElapsedTime  DistanceGroup  \
28459   NaN              0.0       195.0              4               
193274  NaN              0.0       74.0               2               
183449  NaN              0.0       295.0              8               
157662  NaN              0.0       107.0              3               
304586  NaN              0.0       249.0              9               

        CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
28459   76.0          0.0           15.0      0.0            0.0                
193274 NaN           NaN           NaN       NaN            NaN                 
183449 NaN           NaN           NaN       NaN            NaN                 
157662 NaN           NaN           NaN       NaN            NaN                 
304586 NaN           NaN           NaN       NaN            NaN

In [6]:
# For Logistic Regression the data should not have categorical values, convert to dummy codes
# perform one-hot encoding of the categorical data "Carrier"
tmp_df = pd.get_dummies(df_sampled.Carrier,prefix='Carrier')
df_sampled = pd.concat((df_sampled,tmp_df),axis=1) # add back into the dataframe

# perform one-hot encoding of the categorical data "Origin"
tmp_df = pd.get_dummies(df_sampled.Origin,prefix='Origin')
df_sampled = pd.concat((df_sampled,tmp_df),axis=1) # add back into the dataframe

# perform one-hot encoding of the categorical data "Destination"
tmp_df = pd.get_dummies(df_sampled.Dest,prefix='Dest')
df_sampled = pd.concat((df_sampled,tmp_df),axis=1) # add back into the dataframe

# perform one-hot encoding of the categorical data "DepTimeBlk"
tmp_df = pd.get_dummies(df_sampled.DepTimeBlk,prefix='DepTime')
df_sampled = pd.concat((df_sampled,tmp_df),axis=1) # add back into the dataframe

# perform one-hot encoding of the categorical data "ArrTimeBlk"
tmp_df = pd.get_dummies(df_sampled.ArrTimeBlk,prefix='ArrTime')
df_sampled = pd.concat((df_sampled,tmp_df),axis=1) # add back into the dataframe

# perform one-hot encoding of the categorical data "CancellationCodes"
tmp_df = pd.get_dummies(df_sampled.CancellationCode,prefix='CancellationCode')
df_sampled = pd.concat((df_sampled,tmp_df),axis=1) # add back into the dataframe

In [7]:
df_sampled.describe( include = 'all')  # all when mixed

Month     DayofMonth Carrier  Origin    Dest  DepDelayMinutes  \
count   560850.000000  560850.000000  560850  560850  560850  554741.000000     
unique NaN            NaN             12      311     309    NaN                
top    NaN            NaN             WN      ATL     ATL    NaN                
freq   NaN            NaN             130609  38151   38374  NaN                
mean    6.542491       15.733694      NaN     NaN     NaN     12.142479         
std     3.399781       8.772927       NaN     NaN     NaN     39.351163         
min     1.000000       1.000000       NaN     NaN     NaN     0.000000          
25%     4.000000       8.000000       NaN     NaN     NaN     0.000000          
50%     7.000000       16.000000      NaN     NaN     NaN     0.000000          
75%     9.000000       23.000000      NaN     NaN     NaN     6.000000          
max     12.000000      31.000000      NaN     NaN     NaN     1964.000000       

       DepTimeBlk        TaxiOut         TaxiIn  ArrDelayMinutes ArrTimeBlk  \
count   560850     554558.000000  554289.000000  553046.000000    560850      
unique  19        NaN            NaN            NaN               19          
top     0700-0759 NaN            NaN            NaN               1600-1659   
freq    38558     NaN            NaN            NaN               37732       
mean    NaN        16.220159      7.494479       12.164382        NaN         
std     NaN        8.818646       5.826084       39.044046        NaN         
min     NaN        1.000000       1.000000       0.000000         NaN         
25%     NaN        11.000000      4.000000       0.000000         NaN         
50%     NaN        14.000000      6.000000       0.000000         NaN         
75%     NaN        19.000000      9.000000       7.000000         NaN         
max     NaN        174.000000     413.000000     1971.000000      NaN         

            Cancelled CancellationCode       Diverted  ActualElapsedTime  \
count   560850.000000  6339             560850.000000  553046.000000       
unique NaN             4               NaN            NaN                  
top    NaN             B               NaN            NaN                  
freq   NaN             3241            NaN            NaN                  
mean    0.011302       NaN              0.002612       140.402480          
std     0.105711       NaN              0.051042       75.924852           
min     0.000000       NaN              0.000000       15.000000           
25%     0.000000       NaN              0.000000       84.000000           
50%     0.000000       NaN              0.000000       122.000000          
75%     0.000000       NaN              0.000000       173.000000          
max     1.000000       NaN              1.000000       759.000000          

        DistanceGroup  CarrierDelay  WeatherDelay      NASDelay  \
count   560850.000000  98612.000000  98612.000000  98612.000000   
unique NaN            NaN           NaN           NaN             
top    NaN            NaN           NaN           NaN             
freq   NaN            NaN           NaN           NaN             
mean    3.873499       20.013518     2.716748      14.853436      
std     2.439576       57.434797     21.447157     31.327341      
min     1.000000       0.000000      0.000000      0.000000       
25%     2.000000       0.000000      0.000000      0.000000       
50%     3.000000       1.000000      0.000000      2.000000       
75%     5.000000       18.000000     0.000000      19.000000      
max     11.000000      1964.000000   1061.000000   1446.000000    

        SecurityDelay  LateAircraftDelay     Carrier_AA     Carrier_AS  \
count   98612.000000   98612.000000       560850.000000  560850.000000   
unique NaN            NaN                NaN            NaN              
top    NaN            NaN                NaN            NaN              
freq   NaN            NaN                NaN            NaN              
mean  

In [19]:
# Cancelled and diverted are categorical variables, but they appear above as float values.  No need to convert them as 1 equals Cancelled or Diverted.
#df_sampled['Cancelled']=df_sampled['Cancelled'].astype(str)  # realized this is not necessary
#df_sampled['Diverted']=df_sampled['Diverted'].astype(str)  # realized this is not necessary


In [8]:
df_sampled.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 560850 entries, 28459 to 172988
Columns: 695 entries, Month to CancellationCode_D
dtypes: float64(12), int64(3), object(6), uint8(674)
memory usage: 454.6+ MB


#### There are 674 dummy variables in the data to model with, but the columns that were converted to dummy variables still exist.  

###Remove them.

In [9]:
# Remove the original source of the dummy fields, the ones with Characteristic values (see output above to locate them all)
if 'Carrier' in df_sampled:
    del df_sampled['Carrier']
    
# Remove the original source of the dummy fields, the ones with Characteristic values (see output above to locate them all)
if 'Origin' in df_sampled:
    del df_sampled['Origin']
    
# Remove the original source of the dummy fields, the ones with Characteristic values (see output above to locate them all)
if 'Dest' in df_sampled:
    del df_sampled['Dest']

# Remove the original source of the dummy fields, the ones with Characteristic values (see output above to locate them all)
if 'DepTimeBlk' in df_sampled:
    del df_sampled['DepTimeBlk']
    
# Remove the original source of the dummy fields, the ones with Characteristic values (see output above to locate them all)
if 'ArrTimeBlk' in df_sampled:
    del df_sampled['ArrTimeBlk']
    
    # Remove the original source of the dummy fields, the ones with Characteristic values (see output above to locate them all)
if 'CancellationCode' in df_sampled:
    del df_sampled['CancellationCode']
    


In [10]:
df_sampled.describe( include = 'all')  # all when mixed

Month     DayofMonth  DepDelayMinutes        TaxiOut  \
count  560850.000000  560850.000000  554741.000000    554558.000000   
mean   6.542491       15.733694      12.142479        16.220159       
std    3.399781       8.772927       39.351163        8.818646        
min    1.000000       1.000000       0.000000         1.000000        
25%    4.000000       8.000000       0.000000         11.000000       
50%    7.000000       16.000000      0.000000         14.000000       
75%    9.000000       23.000000      6.000000         19.000000       
max    12.000000      31.000000      1964.000000      174.000000      

              TaxiIn  ArrDelayMinutes      Cancelled       Diverted  \
count  554289.000000  553046.000000    560850.000000  560850.000000   
mean   7.494479       12.164382        0.011302       0.002612        
std    5.826084       39.044046        0.105711       0.051042        
min    1.000000       0.000000         0.000000       0.000000        
25%    4.000000       0.000000         0.000000       0.000000        
50%    6.000000       0.000000         0.000000       0.000000        
75%    9.000000       7.000000         0.000000       0.000000        
max    413.000000     1971.000000      1.000000       1.000000        

       ActualElapsedTime  DistanceGroup  CarrierDelay  WeatherDelay  \
count  553046.000000      560850.000000  98612.000000  98612.000000   
mean   140.402480         3.873499       20.013518     2.716748       
std    75.924852          2.439576       57.434797     21.447157      
min    15.000000          1.000000       0.000000      0.000000       
25%    84.000000          2.000000       0.000000      0.000000       
50%    122.000000         3.000000       1.000000      0.000000       
75%    173.000000         5.000000       18.000000     0.000000       
max    759.000000         11.000000      1964.000000   1061.000000    

           NASDelay  SecurityDelay  LateAircraftDelay     Carrier_AA  \
count  98612.000000  98612.000000   98612.000000       560850.000000   
mean   14.853436     0.072800       24.657121          0.161717        
std    31.327341     1.908304       46.578801          0.368191        
min    0.000000      0.000000       0.000000           0.000000        
25%    0.000000      0.000000       0.000000           0.000000        
50%    2.000000      0.000000       3.000000           0.000000        
75%    19.000000     0.000000       30.000000          0.000000        
max    1446.000000   168.000000     1257.000000        1.000000        

          Carrier_AS     Carrier_B6     Carrier_DL     Carrier_EV  \
count  560850.000000  560850.000000  560850.000000  560850.000000   
mean   0.031718       0.050892       0.163978       0.084761        
std    0.175248       0.219778       0.370256       0.278525        
min    0.000000       0.000000       0.000000       0.000000        
25%    0.000000       0.000000       0.000000       0.000000        
50%    0.000000       0.000000       0.000000       0.000000        
75%    0.000000       0.000000       0.000000       0.000000        
max    1.000000       1.000000       1.000000       1.000000        

          Carrier_F9     Carrier_HA     Carrier_NK     Carrier_OO  \
count  560850.000000  560850.000000  560850.000000  560850.000000   
mean   0.016972       0.013633       0.025406       0.108093        
std    0.129168       0.115961       0.157355       0.310498        
min    0.000000       0.000000       0.000000       0.000000        
25%    0.000000       0.000000       0.000000       0.000000        
50%    0.000000       0.000000       0.000000       0.000000        
75%    0.000000       0.000000       0.000000       0.000000        
max    1.000000       1.000000       1.000000       1.000000        

          Carrier_UA     Carrier_VX     Carrier_WN     Origin_ABE  \
count  560850.000000  560850.000000  560850.000000  560850.000000   
mean   0.097589       0.012363       0.232877       0.000419    

##Review of the data - 
#####  Examination of the min and max should make sense, such as Month with 1 and 12, DayofMonth with 1 and 31 and so on.
####  Cancelled and Diverted are min and max of 0 and 1 have a semantic of 1 = True for the characteristic
####  

### We will be modelling the arrival delay in minutes for which we will create a new variable that normalizes the delay by the duration of the flight by creating a new variable ArrivalDelayPerElapsedTime


In [12]:
df_sampled['ArrDelayPerElapsedTime'] = df_sampled.ArrDelayMinutes / df_sampled.ActualElapsedTime
df_sampled.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 560850 entries, 28459 to 172988
Columns: 690 entries, Month to ArrDelayPerElapsedTime
dtypes: float64(13), int64(3), uint8(674)
memory usage: 433.2 MB


In [13]:
df_sampled.head()

Month  DayofMonth  DepDelayMinutes  TaxiOut  TaxiIn  ArrDelayMinutes  \
28459   3      18          76.0             31.0     10.0    91.0              
193274  3      19          0.0              20.0     3.0     0.0               
183449  3      27          0.0              10.0     20.0    0.0               
157662  3      18          0.0              12.0     19.0    0.0               
304586  3      18          7.0              15.0     3.0     0.0               

        Cancelled  Diverted  ActualElapsedTime  DistanceGroup  CarrierDelay  \
28459   0.0        0.0       195.0              4              76.0           
193274  0.0        0.0       74.0               2             NaN             
183449  0.0        0.0       295.0              8             NaN             
157662  0.0        0.0       107.0              3             NaN             
304586  0.0        0.0       249.0              9             NaN             

        WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  Carrier_AA  \
28459   0.0           15.0      0.0            0.0                1            
193274 NaN           NaN       NaN            NaN                 0            
183449 NaN           NaN       NaN            NaN                 0            
157662 NaN           NaN       NaN            NaN                 0            
304586 NaN           NaN       NaN            NaN                 0            

        Carrier_AS  Carrier_B6  Carrier_DL  Carrier_EV  Carrier_F9  \
28459   0           0           0           0           0            
193274  0           0           1           0           0            
183449  0           0           1           0           0            
157662  0           0           1           0           0            
304586  0           0           0           0           0            

        Carrier_HA  Carrier_NK  Carrier_OO  Carrier_UA  Carrier_VX  \
28459   0           0           0           0           0            
193274  0           0           0           0           0            
183449  0           0           0           0           0            
157662  0           0           0           0           0            
304586  0           0           0           0           0            

        Carrier_WN  Origin_ABE  Origin_ABI  Origin_ABQ  Origin_ABR  \
28459   0           0           0           0           0            
193274  0           0           0           0           0            
183449  0           0           0           0           0            
157662  0           0           0           0           0            
304586  1           0           0           0           0            

        Origin_ABY  Origin_ACK  Origin_ACT  Origin_ACV  Origin_ACY  \
28459   0           0           0           0           0            
193274  0           0           0           0           0            
183449  0           0           0           0           0            
157662  0           0           0           0           0            
304586  0           0           0           0           0            

        Origin_ADK  Origin_ADQ  Origin_AEX  Origin_AGS  Origin_AKN  \
28459   0           0           0           0           0            
193274  0           0           0           0           0            
183449  0           0           0           0           0            
157662  0           0           0           0           0            
304586  0           0           0           0           0            

        Origin_ALB  Origin_AMA  Origin_ANC  Origin_APN  Origin_ASE  \
28459   0           0           0           0           0            
193274  0           0           0           0           0            
183449  0           0           0           0           0            
157662  0           0           0           0           0            
304586  0           0           0           0           0            

        Origin_ATL  Origin_ATW  Origin_AUS  O

##  Create a composite indicator for delay, cancellation or Diversion using the formula that if arrival delay is greater than 15 min then count it as 1 (delayed) else as 0 (delayed) and add it to the diverted and cancelled

In [17]:
## create a new column with a default value of 0 to show no problems and 1 if delayed past 15 min, cancelled or  diverted
#df_sampled['OnTime'] = '0'
#df_sampled['OnTime'][df_sampled['Cancelled']>0] = '1'
#df_sampled['OnTime'][df_sampled['Diverted']>0] = '1'
df_sampled['OnTime']=df_sampled['OnTime'].astype(int)  # realized this is not necessary
df_sampled.head(20)

Month  DayofMonth  DepDelayMinutes  TaxiOut  TaxiIn  ArrDelayMinutes  \
28459   3      18          76.0             31.0     10.0    91.0              
193274  3      19          0.0              20.0     3.0     0.0               
183449  3      27          0.0              10.0     20.0    0.0               
157662  3      18          0.0              12.0     19.0    0.0               
304586  3      18          7.0              15.0     3.0     0.0               
420863  3      31          0.0              10.0     6.0     0.0               
159510  3      19          0.0              10.0     7.0     0.0               
52033   3      14          0.0              31.0     36.0    15.0              
222509  3      11          34.0             11.0     24.0    33.0              
461259  3      10          0.0              11.0     12.0    0.0               
15814   3      21          0.0              15.0     2.0     0.0               
46104   3      1           0.0              19.0     5.0     0.0               
135149  3      9           15.0             23.0     7.0     14.0              
474538  3      2           0.0              19.0     8.0     1.0               
175729  3      24          11.0             90.0     6.0     70.0              
240921  3      31          5.0              9.0      4.0     0.0               
134191  3      10          0.0              13.0     9.0     0.0               
126700  3      4           0.0              19.0     3.0     0.0               
16714   3      6           18.0             39.0     11.0    27.0              
9554    3      10          48.0             13.0     7.0     37.0              

        Cancelled  Diverted  ActualElapsedTime  DistanceGroup  CarrierDelay  \
28459   0.0        0.0       195.0              4              76.0           
193274  0.0        0.0       74.0               2             NaN             
183449  0.0        0.0       295.0              8             NaN             
157662  0.0        0.0       107.0              3             NaN             
304586  0.0        0.0       249.0              9             NaN             
420863  0.0        0.0       113.0              3             NaN             
159510  0.0        0.0       166.0              6             NaN             
52033   0.0        0.0       272.0              6              0.0            
222509  0.0        0.0       87.0               2              0.0            
461259  0.0        0.0       252.0              7             NaN             
15814   0.0        0.0       184.0              5             NaN             
46104   0.0        0.0       73.0               1             NaN             
135149  0.0        0.0       312.0              8             NaN             
474538  0.0        0.0       189.0              5             NaN             
175729  0.0        0.0       169.0              3              0.0            
240921  0.0        0.0       144.0              4             NaN             
134191  0.0        0.0       98.0               3             NaN             
126700  0.0        0.0       154.0              5             NaN             
16714   0.0        0.0       231.0              7              0.0            
9554    0.0        0.0       155.0              5              13.0           

        WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  Carrier_AA  \
28459   0.0           15.0      0.0            0.0                1            
193274 NaN           NaN       NaN            NaN                 0            
183449 NaN           NaN       NaN            NaN                 0            
157662 NaN           NaN       NaN            NaN                 0            
304586 NaN           NaN       NaN            NaN                 0            
420863 NaN           NaN       NaN            NaN                 0            
159510 NaN           NaN       NaN            NaN                 0            
52033   0.0           15.0 

In [18]:
df_sampled.describe()

Month     DayofMonth  DepDelayMinutes        TaxiOut  \
count  560850.000000  560850.000000  554741.000000    554558.000000   
mean   6.542491       15.733694      12.142479        16.220159       
std    3.399781       8.772927       39.351163        8.818646        
min    1.000000       1.000000       0.000000         1.000000        
25%    4.000000       8.000000       0.000000         11.000000       
50%    7.000000       16.000000      0.000000         14.000000       
75%    9.000000       23.000000      6.000000         19.000000       
max    12.000000      31.000000      1964.000000      174.000000      

              TaxiIn  ArrDelayMinutes      Cancelled       Diverted  \
count  554289.000000  553046.000000    560850.000000  560850.000000   
mean   7.494479       12.164382        0.011302       0.002612        
std    5.826084       39.044046        0.105711       0.051042        
min    1.000000       0.000000         0.000000       0.000000        
25%    4.000000       0.000000         0.000000       0.000000        
50%    6.000000       0.000000         0.000000       0.000000        
75%    9.000000       7.000000         0.000000       0.000000        
max    413.000000     1971.000000      1.000000       1.000000        

       ActualElapsedTime  DistanceGroup  CarrierDelay  WeatherDelay  \
count  553046.000000      560850.000000  98612.000000  98612.000000   
mean   140.402480         3.873499       20.013518     2.716748       
std    75.924852          2.439576       57.434797     21.447157      
min    15.000000          1.000000       0.000000      0.000000       
25%    84.000000          2.000000       0.000000      0.000000       
50%    122.000000         3.000000       1.000000      0.000000       
75%    173.000000         5.000000       18.000000     0.000000       
max    759.000000         11.000000      1964.000000   1061.000000    

           NASDelay  SecurityDelay  LateAircraftDelay     Carrier_AA  \
count  98612.000000  98612.000000   98612.000000       560850.000000   
mean   14.853436     0.072800       24.657121          0.161717        
std    31.327341     1.908304       46.578801          0.368191        
min    0.000000      0.000000       0.000000           0.000000        
25%    0.000000      0.000000       0.000000           0.000000        
50%    2.000000      0.000000       3.000000           0.000000        
75%    19.000000     0.000000       30.000000          0.000000        
max    1446.000000   168.000000     1257.000000        1.000000        

          Carrier_AS     Carrier_B6     Carrier_DL     Carrier_EV  \
count  560850.000000  560850.000000  560850.000000  560850.000000   
mean   0.031718       0.050892       0.163978       0.084761        
std    0.175248       0.219778       0.370256       0.278525        
min    0.000000       0.000000       0.000000       0.000000        
25%    0.000000       0.000000       0.000000       0.000000        
50%    0.000000       0.000000       0.000000       0.000000        
75%    0.000000       0.000000       0.000000       0.000000        
max    1.000000       1.000000       1.000000       1.000000        

          Carrier_F9     Carrier_HA     Carrier_NK     Carrier_OO  \
count  560850.000000  560850.000000  560850.000000  560850.000000   
mean   0.016972       0.013633       0.025406       0.108093        
std    0.129168       0.115961       0.157355       0.310498        
min    0.000000       0.000000       0.000000       0.000000        
25%    0.000000       0.000000       0.000000       0.000000        
50%    0.000000       0.000000       0.000000       0.000000        
75%    0.000000       0.000000       0.000000       0.000000        
max    1.000000       1.000000       1.000000       1.000000        

          Carrier_UA     Carrier_VX     Carrier_WN     Origin_ABE  \
count  560850.000000  560850.000000  560850.000000  560850.000000   
mean   0.097589       0.012363       0.232877       0.000419    